File: ios/linktrail_flutter/Sources/linktrail_flutter/LinktrailFlutterPlugin.swift
import Flutter
import LinkTrailSDK
import UIKit

/// Flutter plugin wrapping the native `LinkTrailSDK` iOS SDK.
public class LinktrailFlutterPlugin: NSObject, FlutterPlugin, FlutterSceneLifeCycleDelegate {
  fileprivate var onLinkSink: FlutterEventSink?
  fileprivate var onAttributionSink: FlutterEventSink?
  fileprivate var onErrorSink: FlutterEventSink?

  /// On a cold launch the open-URL / Universal-Link callbacks can fire before Dart has called
  /// configure(), so LinkTrail.shared is still nil and the link would be lost. Buffer it here and
  /// replay it once configure() runs.
  private var pendingLaunchURL: URL?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = LinktrailFlutterPlugin()

    let channel = FlutterMethodChannel(name: "linktrail_flutter", binaryMessenger: registrar.messenger())
    registrar.addMethodCallDelegate(instance, channel: channel)

    FlutterEventChannel(name: "linktrail_flutter/onLink", binaryMessenger: registrar.messenger())
      .setStreamHandler(LinkStreamHandler(plugin: instance))
    FlutterEventChannel(name: "linktrail_flutter/onAttribution", binaryMessenger: registrar.messenger())
      .setStreamHandler(AttributionStreamHandler(plugin: instance))
    FlutterEventChannel(name: "linktrail_flutter/onError", binaryMessenger: registrar.messenger())
      .setStreamHandler(ErrorStreamHandler(plugin: instance))

    // Auto-capture Universal Links / custom-scheme opens without any AppDelegate boilerplate.
    // Apps on the classic UIApplicationDelegate lifecycle deliver these to addApplicationDelegate;
    // apps on the iOS 13+ UIScene lifecycle (the default in recent Flutter templates, required on
    // iOS 26) deliver them to addSceneDelegate instead. Register for both so links are captured
    // regardless of which lifecycle the host app uses.
    registrar.addApplicationDelegate(instance)
    registrar.addSceneDelegate(instance)
  }

  /// (Re)registers the native callback hooks on the current `LinkTrail.shared` instance.
  fileprivate func attachHooks() {
    guard let sdk = LinkTrail.shared else { return }
    sdk.onLink { [weak self] link, source in
      self?.onLinkSink?(["link": link.toMap(), "source": source.dartName])
    }
    sdk.onAttribution { [weak self] attribution in
      self?.onAttributionSink?(attribution.toMap())
    }
    sdk.onError { [weak self] error in
      self?.onErrorSink?(toErrorMap(error))
    }
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let args = call.arguments as? [String: Any]

    switch call.method {
    case "configure":
      guard let apiKey = args?["apiKey"] as? String else {
        result(FlutterError(code: "missingApiKey", message: "An API key is required to configure LinkTrail.", details: nil))
        return
      }
      do {
        _ = try LinkTrail.configure(apiKey: apiKey, options: Self.parseOptions(args?["options"] as? [String: Any]))
        attachHooks()
        // Replay a link that launched the app cold, before the SDK was configured.
        if let url = pendingLaunchURL {
          pendingLaunchURL = nil
          LinkTrail.shared?.handleDeepLink(url)
        }
        result(nil)
      } catch {
        result(toFlutterError(error))
      }

    case "handleDeepLink":
      guard let urlString = args?["url"] as? String, let url = URL(string: urlString), let sdk = LinkTrail.shared else {
        result(false)
        return
      }
      result(sdk.handleDeepLink(url))

    case "trackInstall":
      guard let sdk = LinkTrail.shared else {
        result(FlutterError.notConfigured)
        return
      }
      let force = args?["force"] as? Bool ?? false
      Task {
        do {
          result(try await sdk.trackInstall(force: force).toMap())
        } catch {
          result(toFlutterError(error))
        }
      }

    case "trackEvent":
      guard let sdk = LinkTrail.shared else {
        result(FlutterError.notConfigured)
        return
      }
      guard let name = args?["name"] as? String else {
        result(FlutterError(code: "invalidArgument", message: "trackEvent requires a name.", details: nil))
        return
      }
      let value = args?["value"] as? Double
      let currency = args?["currency"] as? String
      Task {
        do {
          result(try await sdk.trackEvent(name: name, value: value, currency: currency).toMap())
        } catch {
          result(toFlutterError(error))
        }
      }

    case "getLastAttribution":
      result(LinkTrail.shared?.lastAttribution?.toMap())

    case "getLastDeepLink":
      result(LinkTrail.shared?.lastDeepLink?.toMap())

    case "requestTrackingAuthorization":
      guard let sdk = LinkTrail.shared else {
        result(false)
        return
      }
      Task { result(await sdk.requestTrackingAuthorization()) }

    case "registerForSKAdAttribution":
      LinkTrail.shared?.registerForSKAdAttribution()
      result(nil)

    case "updateConversionValue":
      let value = args?["value"] as? Int ?? 0
      let coarse = (args?["coarseValue"] as? String).flatMap(LinkTrailCoarseConversionValue.init(dartName:))
      LinkTrail.shared?.updateConversionValue(value, coarseValue: coarse)
      result(nil)

    case "resetForTesting":
      LinkTrail.resetForTesting()
      result(nil)

    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private static func parseOptions(_ map: [String: Any]?) -> LinkTrailOptions {
    guard let map else { return LinkTrailOptions() }
    let retryMap = map["retryPolicy"] as? [String: Any]
    let logLevel = (map["logLevel"] as? String).flatMap(LinkTrailLogLevel.init(dartName:)) ?? .info
    return LinkTrailOptions(
      logEnabled: map["logEnabled"] as? Bool ?? false,
      logLevel: logLevel,
      requestTimeout: ((map["requestTimeoutMillis"] as? NSNumber)?.doubleValue ?? 15_000) / 1000,
      retryPolicy: LinkTrailRetryPolicy(
        maxAttempts: (retryMap?["maxAttempts"] as? NSNumber)?.intValue ?? 3,
        baseDelay: ((retryMap?["baseDelayMillis"] as? NSNumber)?.doubleValue ?? 500) / 1000,
        maxDelay: ((retryMap?["maxDelayMillis"] as? NSNumber)?.doubleValue ?? 8_000) / 1000
      ),
      linkDomains: map["linkDomains"] as? [String] ?? [],
      autoTrackInstall: map["autoTrackInstall"] as? Bool ?? true
    )
  }

  // -- Auto-capture (classic UIApplicationDelegate lifecycle). --

  public func application(_ application: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey: Any] = [:]) -> Bool {
    handleOrBufferLaunchURL(url)
  }

  public func application(
    _ application: UIApplication,
    continue userActivity: NSUserActivity,
    restorationHandler: @escaping ([UIUserActivityRestoring]?) -> Void
  ) -> Bool {
    guard userActivity.activityType == NSUserActivityTypeBrowsingWeb, let url = userActivity.webpageURL else {
      return false
    }
    return handleOrBufferLaunchURL(url)
  }

  // -- Auto-capture (iOS 13+ UIScene lifecycle — the path used on iOS 26). --

  /// Cold start: the app was launched from a killed state by a link, delivered in the scene's
  /// connection options. Fires before Dart configure(), so the URL is buffered and replayed.
  public func scene(
    _ scene: UIScene,
    willConnectTo session: UISceneSession,
    options connectionOptions: UIScene.ConnectionOptions?
  ) -> Bool {
    var handled = false
    for context in connectionOptions?.urlContexts ?? [] {
      handled = handleOrBufferLaunchURL(context.url) || handled
    }
    for activity in connectionOptions?.userActivities ?? [] where activity.activityType == NSUserActivityTypeBrowsingWeb {
      if let url = activity.webpageURL { handled = handleOrBufferLaunchURL(url) || handled }
    }
    return handled
  }

  /// Warm: a custom-scheme link opened while the app was already running.
  public func scene(_ scene: UIScene, openURLContexts URLContexts: Set<UIOpenURLContext>) -> Bool {
    var handled = false
    for context in URLContexts {
      handled = handleOrBufferLaunchURL(context.url) || handled
    }
    return handled
  }

  /// Warm: a Universal Link opened while the app was already running.
  public func scene(_ scene: UIScene, continue userActivity: NSUserActivity) -> Bool {
    guard userActivity.activityType == NSUserActivityTypeBrowsingWeb, let url = userActivity.webpageURL else {
      return false
    }
    return handleOrBufferLaunchURL(url)
  }

  /// Handle the URL now if the SDK is configured, otherwise buffer it for replay after configure().
  @discardableResult
  private func handleOrBufferLaunchURL(_ url: URL) -> Bool {
    if let sdk = LinkTrail.shared {
      return sdk.handleDeepLink(url)
    }
    pendingLaunchURL = url
    return true
  }
}

private final class LinkStreamHandler: NSObject, FlutterStreamHandler {
  private weak var plugin: LinktrailFlutterPlugin?
  init(plugin: LinktrailFlutterPlugin) { self.plugin = plugin }

  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    plugin?.onLinkSink = events
    plugin?.attachHooks()
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    plugin?.onLinkSink = nil
    return nil
  }
}

private final class AttributionStreamHandler: NSObject, FlutterStreamHandler {
  private weak var plugin: LinktrailFlutterPlugin?
  init(plugin: LinktrailFlutterPlugin) { self.plugin = plugin }

  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    plugin?.onAttributionSink = events
    plugin?.attachHooks()
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    plugin?.onAttributionSink = nil
    return nil
  }
}

private final class ErrorStreamHandler: NSObject, FlutterStreamHandler {
  private weak var plugin: LinktrailFlutterPlugin?
  init(plugin: LinktrailFlutterPlugin) { self.plugin = plugin }

  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    plugin?.onErrorSink = events
    plugin?.attachHooks()
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    plugin?.onErrorSink = nil
    return nil
  }
}

extension FlutterError {
  static var notConfigured: FlutterError {
    FlutterError(code: "notConfigured", message: "LinkTrail.configure() must be called before this method.", details: nil)
  }
}

extension LinkTrailDeepLink {
  fileprivate func toMap() -> [String: Any?] {
    [
      "slug": slug,
      "url": url,
      "deepLinkPath": deepLinkPath,
      "iosUrl": iosURL,
      "androidUrl": androidURL,
      "fallbackUrl": fallbackURL,
      "campaign": campaign,
      "channel": channel,
      "utm": utm,
      "customData": customData,
    ]
  }
}

extension LinkTrailAttribution {
  fileprivate func toMap() -> [String: Any?] {
    ["id": id, "attributed": attributed, "deepLink": deepLink?.toMap()]
  }
}

extension LinkTrailEventResult {
  fileprivate func toMap() -> [String: Any?] {
    ["id": id, "attributed": attributed]
  }
}

extension LinkTrailLinkSource {
  fileprivate var dartName: String {
    switch self {
    case .deferred: return "deferred"
    case .reengagement: return "reengagement"
    }
  }
}

extension LinkTrailLogLevel {
  fileprivate init?(dartName: String) {
    switch dartName {
    case "debug": self = .debug
    case "info": self = .info
    case "warning": self = .warning
    case "error": self = .error
    case "none": self = .none
    default: return nil
    }
  }
}

extension LinkTrailCoarseConversionValue {
  fileprivate init?(dartName: String) {
    switch dartName {
    case "low": self = .low
    case "medium": self = .medium
    case "high": self = .high
    default: return nil
    }
  }
}

/// Maps a caught native error to a `FlutterError` whose `code` matches the Dart `LinkTrailException` codes.
///
/// A free function rather than an `Error` extension: pattern-matching concrete `LinkTrailError`
/// cases against `Self` inside a protocol extension doesn't type-check (`Self` isn't statically
/// known to be `LinkTrailError`), but matching against a plain `Error`-typed parameter does.
private func toFlutterError(_ error: Error) -> FlutterError {
  let (code, details): (String, [String: Any?]) =
    switch error {
    case LinkTrailError.invalidURL(let url): ("invalidUrl", ["url": url])
    case LinkTrailError.transport(let message): ("transport", ["message": message])
    case LinkTrailError.server(let statusCode, let body): ("server", ["statusCode": statusCode, "body": body])
    case LinkTrailError.decoding(let message): ("decoding", ["message": message])
    case LinkTrailError.emptyResponse: ("emptyResponse", [:])
    case LinkTrailError.notALinkTrailURL: ("notALinkTrailUrl", [:])
    case LinkTrailError.missingAPIKey: ("missingApiKey", [:])
    case LinkTrailError.invalidAPIKey: ("invalidApiKey", [:])
    default: ("unknown", [:])
    }
  return FlutterError(code: code, message: error.localizedDescription, details: details)
}

private func toErrorMap(_ error: Error) -> [String: Any?] {
  let flutterError = toFlutterError(error)
  return ["code": flutterError.code, "message": flutterError.message, "details": flutterError.details]
}

File: android/src/main/kotlin/io/linktrail/linktrail_flutter/LinktrailFlutterPlugin.kt
package io.linktrail.linktrail_flutter

import android.content.Context
import android.content.Intent
import android.net.Uri
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry
import io.linktrail.LinkTrail
import io.linktrail.LinkTrailError
import io.linktrail.LinkTrailLogLevel
import io.linktrail.LinkTrailOptions
import io.linktrail.LinkTrailRetryPolicy
import io.linktrail.model.LinkTrailAttribution
import io.linktrail.model.LinkTrailDeepLink
import io.linktrail.model.LinkTrailEventResult
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

/** Flutter plugin wrapping the native `io.linktrail:sdk` Android SDK. */
class LinktrailFlutterPlugin :
    FlutterPlugin,
    MethodCallHandler,
    ActivityAware,
    PluginRegistry.NewIntentListener {
    private lateinit var context: Context
    private lateinit var channel: MethodChannel
    private lateinit var onLinkChannel: EventChannel
    private lateinit var onAttributionChannel: EventChannel
    private lateinit var onErrorChannel: EventChannel

    private var onLinkSink: EventChannel.EventSink? = null
    private var onAttributionSink: EventChannel.EventSink? = null
    private var onErrorSink: EventChannel.EventSink? = null

    private var activityBinding: ActivityPluginBinding? = null
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    // On a cold start the launch intent is read (in onAttachedToActivity) before Dart has called
    // configure(), so LinkTrail.shared is still null and the link would be lost. Buffer it here and
    // replay it once configure() runs. Warm starts (app already configured) handle links directly.
    private var pendingLaunchUri: Uri? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        val messenger: BinaryMessenger = binding.binaryMessenger

        channel = MethodChannel(messenger, "linktrail_flutter")
        channel.setMethodCallHandler(this)

        onLinkChannel = EventChannel(messenger, "linktrail_flutter/onLink")
        onLinkChannel.setStreamHandler(
            object : EventChannel.StreamHandler {
                override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
                    onLinkSink = events
                    attachHooks()
                }

                override fun onCancel(arguments: Any?) {
                    onLinkSink = null
                }
            },
        )

        onAttributionChannel = EventChannel(messenger, "linktrail_flutter/onAttribution")
        onAttributionChannel.setStreamHandler(
            object : EventChannel.StreamHandler {
                override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
                    onAttributionSink = events
                    attachHooks()
                }

                override fun onCancel(arguments: Any?) {
                    onAttributionSink = null
                }
            },
        )

        onErrorChannel = EventChannel(messenger, "linktrail_flutter/onError")
        onErrorChannel.setStreamHandler(
            object : EventChannel.StreamHandler {
                override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
                    onErrorSink = events
                    attachHooks()
                }

                override fun onCancel(arguments: Any?) {
                    onErrorSink = null
                }
            },
        )
    }

    /** (Re)registers the native callback hooks on the current `LinkTrail.shared` instance. */
    private fun attachHooks() {
        val sdk = LinkTrail.shared ?: return
        sdk.onLink { link, source ->
            onLinkSink?.success(mapOf("link" to link.toMap(), "source" to source.name.lowercase()))
        }
        sdk.onAttribution { attribution -> onAttributionSink?.success(attribution.toMap()) }
        sdk.onError { throwable -> onErrorSink?.success(throwable.toErrorMap()) }
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "configure" -> {
                val apiKey = call.argument<String>("apiKey")
                if (apiKey == null) {
                    result.error("missingApiKey", "An API key is required to configure LinkTrail.", null)
                    return
                }
                try {
                    @Suppress("UNCHECKED_CAST")
                    val optionsMap = call.argument<Map<String, Any?>>("options")
                    LinkTrail.configure(context, apiKey, parseOptions(optionsMap))
                    attachHooks()
                    // Replay a deep link that launched the app cold, before the SDK was configured.
                    pendingLaunchUri?.let { uri ->
                        pendingLaunchUri = null
                        LinkTrail.shared?.handleDeepLink(uri)
                    }
                    result.success(null)
                } catch (e: Throwable) {
                    result.error(e)
                }
            }

            "handleDeepLink" -> {
                val urlString = call.argument<String>("url")
                val sdk = LinkTrail.shared
                if (urlString == null || sdk == null) {
                    result.success(false)
                    return
                }
                result.success(sdk.handleDeepLink(Uri.parse(urlString)))
            }

            "trackInstall" -> {
                val force = call.argument<Boolean>("force") ?: false
                val sdk = LinkTrail.shared ?: return result.errorNotConfigured()
                scope.launch {
                    try {
                        result.success(sdk.trackInstallAsync(force).toMap())
                    } catch (e: Throwable) {
                        result.error(e)
                    }
                }
            }

            "trackEvent" -> {
                val name = call.argument<String>("name")
                val value = call.argument<Double>("value")
                val currency = call.argument<String>("currency")
                val sdk = LinkTrail.shared ?: return result.errorNotConfigured()
                if (name == null) {
                    result.error("invalidArgument", "trackEvent requires a name.", null)
                    return
                }
                scope.launch {
                    try {
                        result.success(sdk.trackEventAsync(name, value, currency).toMap())
                    } catch (e: Throwable) {
                        result.error(e)
                    }
                }
            }

            "getLastAttribution" -> result.success(LinkTrail.shared?.lastAttribution?.toMap())
            "getLastDeepLink" -> result.success(LinkTrail.shared?.lastDeepLink?.toMap())

            // ATT / SKAdNetwork are iOS-only concepts; no-ops on Android.
            "requestTrackingAuthorization" -> result.success(true)
            "registerForSKAdAttribution" -> result.success(null)
            "updateConversionValue" -> result.success(null)

            "resetForTesting" -> {
                LinkTrail.resetForTesting(context)
                result.success(null)
            }

            else -> result.notImplemented()
        }
    }

    private fun Result.errorNotConfigured() {
        error("notConfigured", "LinkTrail.configure() must be called before this method.", null)
    }

    private fun Result.error(e: Throwable) {
        val map = e.toErrorMap()
        error(map["code"] as String, map["message"] as String?, map["details"])
    }

    private fun parseOptions(map: Map<String, Any?>?): LinkTrailOptions {
        if (map == null) return LinkTrailOptions()
        @Suppress("UNCHECKED_CAST")
        val retryMap = map["retryPolicy"] as? Map<String, Any?>
        @Suppress("UNCHECKED_CAST")
        val linkDomains = (map["linkDomains"] as? List<String>) ?: emptyList()
        return LinkTrailOptions(
            logEnabled = map["logEnabled"] as? Boolean ?: false,
            logLevel = LinkTrailLogLevel.valueOf((map["logLevel"] as? String ?: "info").uppercase()),
            requestTimeoutMillis = (map["requestTimeoutMillis"] as? Number)?.toLong() ?: 15_000L,
            retryPolicy =
                LinkTrailRetryPolicy(
                    maxAttempts = (retryMap?.get("maxAttempts") as? Number)?.toInt() ?: 3,
                    baseDelayMillis = (retryMap?.get("baseDelayMillis") as? Number)?.toLong() ?: 500L,
                    maxDelayMillis = (retryMap?.get("maxDelayMillis") as? Number)?.toLong() ?: 8_000L,
                ),
            linkDomains = linkDomains,
            autoTrackInstall = map["autoTrackInstall"] as? Boolean ?: true,
        )
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        onLinkChannel.setStreamHandler(null)
        onAttributionChannel.setStreamHandler(null)
        onErrorChannel.setStreamHandler(null)
    }

    // -- ActivityAware: auto-capture incoming deep links, no consumer boilerplate required. --

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activityBinding = binding
        binding.addOnNewIntentListener(this)
        binding.activity.intent?.data?.let(::handleOrBufferLaunchUri)
    }

    /** Handle the launch URI now if the SDK is configured, otherwise buffer it for replay after configure(). */
    private fun handleOrBufferLaunchUri(uri: Uri) {
        val sdk = LinkTrail.shared
        if (sdk != null) sdk.handleDeepLink(uri) else pendingLaunchUri = uri
    }

    override fun onDetachedFromActivityForConfigChanges() = detachActivity()

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) = onAttachedToActivity(binding)

    override fun onDetachedFromActivity() = detachActivity()

    private fun detachActivity() {
        activityBinding?.removeOnNewIntentListener(this)
        activityBinding = null
    }

    override fun onNewIntent(intent: Intent): Boolean {
        intent.data?.let { uri -> LinkTrail.shared?.handleDeepLink(uri) }
        return false
    }
}

private fun LinkTrailDeepLink.toMap(): Map<String, Any?> =
    mapOf(
        "slug" to slug,
        "url" to url,
        "deepLinkPath" to deepLinkPath,
        "iosUrl" to iosUrl,
        "androidUrl" to androidUrl,
        "fallbackUrl" to fallbackUrl,
        "campaign" to campaign,
        "channel" to channel,
        "utm" to utm,
        "customData" to customData,
    )

private fun LinkTrailAttribution.toMap(): Map<String, Any?> =
    mapOf("id" to id, "attributed" to attributed, "deepLink" to deepLink?.toMap())

private fun LinkTrailEventResult.toMap(): Map<String, Any?> = mapOf("id" to id, "attributed" to attributed)

/** Maps a caught native error to `{code, message, details}`, matching the Dart `LinkTrailException` codes. */
private fun Throwable.toErrorMap(): Map<String, Any?> {
    val noDetails = emptyMap<String, Any?>()
    val (code, details) =
        when (this) {
            is LinkTrailError.InvalidUrl -> "invalidUrl" to mapOf<String, Any?>("url" to url)
            is LinkTrailError.Transport -> "transport" to noDetails
            is LinkTrailError.Server -> "server" to mapOf<String, Any?>("statusCode" to statusCode, "body" to body)
            is LinkTrailError.Decoding -> "decoding" to noDetails
            is LinkTrailError.EmptyResponse -> "emptyResponse" to noDetails
            is LinkTrailError.NotALinkTrailUrl -> "notALinkTrailUrl" to noDetails
            is LinkTrailError.MissingApiKey -> "missingApiKey" to noDetails
            is LinkTrailError.InvalidApiKey -> "invalidApiKey" to noDetails
            else -> "unknown" to noDetails
        }
    return mapOf("code" to code, "message" to (message ?: code), "details" to details)
}
